### ISSUE TITLE
HttpURLConnection in Android working fine, but OkHttp gives “Network is unreachable” on Kitkat

### ISSUE THREAD

**Issue by github-second** [square/okhttp#4378]
I have a problem with this particular site: https://tastedive.com/read/api

If I do an HTTP request with HttpURLConnection, I get a normal HTML response (on Android this code needs to be in a separate thread and also put all the necessary tries and catches):

        StringBuilder result = new StringBuilder();
        URL url = new URL("https://tastedive.com/read/api");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        String line;
        while ((line = rd.readLine()) != null) {
            result.append(line);
        }
        rd.close();
        System.out.println(result.toString());  // shows normal HTML response

But if I do it with OkHttp, with this code...

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url("https://tastedive.com/read/api")
                .build();
        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(okhttp3.Call call, IOException e) {
                Log.d("MY", "failure", e);
            }

            @Override
            public void onResponse(okhttp3.Call call, okhttp3.Response response) throws IOException {
                if (!response.isSuccessful()) {
                    throw new IOException("Unexpected code " + response);
                } else {
                    System.out.println(response.body().string());
                }
            }
        });

...I get this error in onFailure:

       java.net.ConnectException: Failed to connect to tastedive.com/2606:4700:30::681c:3a5:443
           at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:242)
           at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:160)
           at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
           at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
           at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
           at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
           at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
           at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
           at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
           at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
           at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
           at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
           at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
           at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
           at java.lang.Thread.run(Thread.java:841)
        Caused by: java.net.ConnectException: failed to connect to tastedive.com/2606:4700:30::681c:3a5 (port 443) after 10000ms: isConnected failed: ENETUNREACH (Network is unreachable)
           at libcore.io.IoBridge.isConnected(IoBridge.java:223)
           at libcore.io.IoBridge.connectErrno(IoBridge.java:161)
           at libcore.io.IoBridge.connect(IoBridge.java:112)
           at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192)
           at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:459)
           at java.net.Socket.connect(Socket.java:843)
           at okhttp3.internal.platform.AndroidPlatform.connectSocket(AndroidPlatform.java:71)
           at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:240)
           at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:160) 
           at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257) 
           at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) 
           at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) 
           at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200) 
           at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147) 
           at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) 
           at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
           at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
           at java.lang.Thread.run(Thread.java:841) 
        Caused by: libcore.io.ErrnoException: isConnected failed: ENETUNREACH (Network is unreachable)
           at libcore.io.IoBridge.isConnected(IoBridge.java:208)
           at libcore.io.IoBridge.connectErrno(IoBridge.java:161) 
           at libcore.io.IoBridge.connect(IoBridge.java:112) 
           at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192) 
           at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:459) 
           at java.net.Socket.connect(Socket.java:843) 
           at okhttp3.internal.platform.AndroidPlatform.connectSocket(AndroidPlatform.java:71) 
           at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:240) 
           at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:160) 
           at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257) 
           at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) 
           at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) 
           at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) 
           at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) 
           at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200) 
           at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147) 
           at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) 
           at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
           at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
           at java.lang.Thread.run(Thread.java:841) 

This OkHttp that I'm using is from Retrofit 2.4.0. 
Happens on both the emulator and the real device, but only on Kitkat (API 19). Also note that the exception is thrown immediately, not only after 10000 ms.

**Comment by github-second** [comment]
I solved it! The real device was actually getting Javax.net.ssl.SSLHandshakeException, while the emulator was getting java.net.ConnectException. This StackOverflow discussion describes that there's a bug around TLS for pre-Lollipop devices. 
https://stackoverflow.com/questions/29916962/javax-net-ssl-sslhandshakeexception-javax-net-ssl-sslprotocolexception-ssl-han

So now I use this code from this answer: https://stackoverflow.com/a/50640113/9702500, it solves both the emulator exception and the device exception:
```
ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS)
                .supportsTlsExtensions(true)
                .tlsVersions(TlsVersion.TLS_1_2, TlsVersion.TLS_1_1, TlsVersion.TLS_1_0)
                .cipherSuites(
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_DSS_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA)
                .build();
        OkHttpClient client = new OkHttpClient.Builder()
                .connectionSpecs(Collections.singletonList(spec))
                .build();
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://tastedive.com")
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();
```

**Comment by StainlessStlRat** [comment]
Shouldn't okhttp3 apply this by default then? At least for pre-23 devices?

**Comment by blackfatpig** [comment]
Thank you so much! Your code save my life...

**Comment by yschimke** [comment]
> Shouldn't okhttp3 apply this by default then? At least for pre-23 devices?

I'll add a repro to the Android test. So we can easily confirm.

**Comment by sandeep7788** [comment]
error -> 
            HTTP FAILED: java.net.UnknownServiceException: CLEARTEXT communication not enabled for client

after user this 
                     ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS)
                    .supportsTlsExtensions(true)
                   .tlsVersions(TlsVersion.TLS_1_2, TlsVersion.TLS_1_1, TlsVersion.TLS_1_0)
                   .cipherSuites(
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_DSS_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA)
                .build();




How can i solve this.

















Full Class Code -











package com.mdppractice.helper

import android.util.Log
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.*
import java.util.concurrent.TimeUnit


object ServiceGenerator {
    private const val TAG = "@@ServiceGenerator"

    //    private static final String BASE_URL = "https://jsonplaceholder.typicode.com/";
    const val HEADER_CACHE_CONTROL = "Cache-Control"
    const val HEADER_PRAGMA = "Pragma"
    var instance: ServiceGenerator? = null
        get() {
            if (field == null) {
                field = com.mdppractice.helper.ServiceGenerator
            }
            return field
        }
        private set
    private const val cacheSize = (20 * 1024 * 1024).toLong()
    private fun retrofit(): Retrofit {
        return Retrofit.Builder()
            .baseUrl(ApiContants.PREF_base_url)
            .client(okHttpClient())
            .addConverterFactory(GsonConverterFactory.create())
            .build()


    }



    var spec = ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS)
        .supportsTlsExtensions(true)
        .tlsVersions(TlsVersion.TLS_1_2, TlsVersion.TLS_1_1, TlsVersion.TLS_1_0)
        .cipherSuites(
            CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
            CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
            CipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
            CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
            CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
            CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
            CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
            CipherSuite.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
            CipherSuite.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
            CipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
            CipherSuite.TLS_DHE_DSS_WITH_AES_128_CBC_SHA,
            CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA,
            )
        .build()

    private fun okHttpClient(): OkHttpClient {

        val specs = listOf(ConnectionSpec.CLEARTEXT, ConnectionSpec.MODERN_TLS)

        return OkHttpClient.Builder()
            .connectionSpecs(Collections.singletonList(spec))
//            .connectionSpecs(specs)
            .cache(cache())
            .connectTimeout(60, TimeUnit.SECONDS)//1
            .writeTimeout(60, TimeUnit.SECONDS)//2
            .readTimeout(60, TimeUnit.SECONDS)//3
            .addInterceptor(httpLoggingInterceptor()) // used if network off OR on
            .addNetworkInterceptor(networkInterceptor()) // only used when network is on
            .addInterceptor(offlineInterceptor())
            .build()


    }

    private fun cache(): Cache {
        return Cache(File(MdpPratice.getInstance1().cacheDir, "someIdentifier"), cacheSize)
    }

    /**
     * This interceptor will be called both if the network is available and if the network is not available
     * @return
     */
    private fun offlineInterceptor(): Interceptor {
        return Interceptor { chain ->
            Log.d(TAG, "offline interceptor: called.")
            var request = chain.request()

            // prevent caching when network is on. For that we use the "networkInterceptor"
            if (!MdpPratice.hasNetwork()) {
                val cacheControl = CacheControl.Builder()
                    .maxStale(7, TimeUnit.DAYS)
                    .build()
                request = request.newBuilder()
                    .addHeader("Content-Type", "application/json")
                    .removeHeader(HEADER_PRAGMA)
                    .removeHeader(HEADER_CACHE_CONTROL)
                    .cacheControl(cacheControl)
                    .build()
            }
            chain.proceed(request)
        }
    }

    private fun networkInterceptor(): Interceptor {
        return Interceptor { chain ->
            Log.d(TAG, "network interceptor: called.")
            val response = chain.proceed(chain.request())
            val cacheControl = CacheControl.Builder()
                .maxAge(5, TimeUnit.SECONDS)
                .build()
            response.newBuilder()
                .removeHeader(HEADER_PRAGMA)
                .removeHeader(HEADER_CACHE_CONTROL)
                .header(HEADER_CACHE_CONTROL, cacheControl.toString())
                .build()
        }
    }

    private fun httpLoggingInterceptor(): HttpLoggingInterceptor {
        val httpLoggingInterceptor = HttpLoggingInterceptor { message ->
            Log.d(TAG,
                "log: http log: $message")
        }
        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        return httpLoggingInterceptor
    }

    val api: ApiInterface
        get() = retrofit().create(ApiInterface::class.java)
}

**Comment by yschimke** [comment]
@sandeep7788 please ask question like this on stackoverflow, not piggybacking on another issue.  Thanks.
